src/unix/lwt_unix.cppo.ml

```diff
@@ -1224,46 +1224,28 @@ let access name mode =
    | Operations on file descriptors                                  |
    +-----------------------------------------------------------------+ *)
 
-let dup ch =
-  check_descriptor ch;
-  let fd = Unix.dup ch.fd in
-  {
-    fd = fd;
+let dup ch1 =
+  check_descriptor ch1;
+  let ch2 = {
+    fd = Unix.dup ch1.fd;
+    set_flags = true;                 (* overridden *)
+    blocking = lazy(Lwt.return_true); (* overridden *)
     state = Opened;
-    set_flags = ch.set_flags;
-    blocking =
-      if ch.set_flags then
-        lazy(Lazy.force ch.blocking >>= function
-        | true ->
-          Unix.clear_nonblock fd;
-          Lwt.return_true
-        | false ->
-          Unix.set_nonblock fd;
-          Lwt.return_false)
-      else
-        ch.blocking;
     event_readable = None;
     event_writable = None;
     hooks_readable = Lwt_sequence.create ();
     hooks_writable = Lwt_sequence.create ();
-  }
+  } in
+  set_blocking ~set_flags:ch1.set_flags ch2 ch1.set_flags;
+  ch2
 
 let dup2 ch1 ch2 =
   check_descriptor ch1;
   Unix.dup2 ch1.fd ch2.fd;
-  ch2.set_flags <- ch1.set_flags;
```

Comment:
This line is not necessary for the establishment of the `blocking` field anymore. However, removing this line also means that `ch2` is in a different state when the function returns. (In the previous behaviour, it was set explicitly now it isn't.)

I'm not sure it can cause issues but maybe? Consider the following case:

```
let cha = … in (* some ch, whatever *)
let chb = … in (* some other ch *)
let dup2ed = dup2 cha chb in (* chb's `set_flag` is unaffected by this *)
… (* some stuff here *)
let duped = dup dup2ed in (* possible issue here *)
…
```

The possible issue comes from the fact that `dup` reads `dup2ed.set_flags` which is set lazily when `dup2ed`'s `blocking field is forced. And so the value of the `set_flags` and `blocking` fields for `duped`, depends on whether `some stuff here` forces this field or not.


----------------------

Am I missing something? Is there a guarantee that the field is forced? Is there another piece of the control-flow I haven't seen?